Nordic (s-file) integration.

calum-chamberlain opened the pull request:
This pull request relates to the previous pull-request [1316](https://github.com/obspy/obspy/pull/1316).  I have moved across the [EQcorrscan](https://github.com/eqcorrscan/EQcorrscan) sfile_util submodule to here which has quite a bit (but not all) read/write support for nordic sfiles.

What it does:
- Read and write location (origin) information to obspy.core.event.Event
- Read and write magnitude info (including relevant picks);
- Read spectral info from nordic files, but this isn't stored in obspy structure at the moment, just in a dict
- Read in pick information and uncertainties, possibly not ideal mapping for uncertainties, welcome suggestions

What it doesn't do:
- Read or write in focal mechanism information, I don't have a useful example at the moment, but it should be straight-forward to add, I just wanted to get this going
- Read or write moment tensor info, again, no example to play with.
- Read or write Hyp error assesments

I have included my tests, which pass locally, shall see how they go on travis and appveyor (they were working for eqcorrscan before), and see how coverage goes.

Still to do:
- [x] Test that the plugin is found properly and that it plays nicely with read_events
- [x] Add to CONTRIBUTING and CHANGELOG
- [x]  Add test for plugin for read_events
- [x] Remove NaN/999 mappings;
- [x] Remove temporary file usage (use io objects);
- [x] Minor comments from @krischer;
- [x] Check docs;
- [x] stationtoseisan map to `Inventory.write`?;
- [x] Remove IOError usage, replace with custom error;
- [x] Do not use default filename for writing, and set default for overwrite to True.

krischer commented on obspy/CONTRIBUTORS.txt:
```diff
@@ -67,4 +67,5 @@ Andrew Walker
-Andrew Winkelman
\ No newline at end of file
+Andrew Winkelman
+Calum Chamberlain
```
The list is sorted by last name - can you put yourself in the correct place?

krischer commented on obspy/core/util/base.py:
```diff
@@ -41,7 +41,7 @@
                    'io.shapefile', 'io.seiscomp', 'io.stationtxt',
                    'io.stationxml', 'io.wav', 'io.xseed', 'io.y', 'io.zmap',
-                   'realtime', 'signal', 'taup']
+                   'io.nordic', 'realtime', 'signal', 'taup']
```
This list is also sorted alphabetically.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1338 @@
+# -*- coding: utf-8 -*-
+"""
+NonLinLoc file format support for ObsPy
```
Leftover from the template I guess.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1338 @@
+:license:
+    GNU Lesser General Public License, Version 3
+    (https://www.gnu.org/copyleft/lesser.html)
+We have not implemented any handling of focal mechanism solutions between
```
Missing newline in front.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1338 @@
+.. note:: Pick time-residuals are handled in event.origins[0].arrivals, with \
+    the arrival.pick_id linking the arrival (which contain calculated \
+    information) with the pick.resource_id (where the pick contains only \
+    physical measured information).
```
No need for the backslashes. It should be possible to just format it as

```rst
.. note::

    TEXT OF NOTE

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1338 @@
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+from future.builtins import *  # NOQA @UnusedWildImport
+from six import string_types
```
`six` is not an ObsPy dependency so you cannot use it. You can replace it with

```python
from future.utils import PY2

if PY2:
    string_types = basestring
else:
    string_types = str
```

QuLogic replied:
Generally, you should not need `string_types`; you should be explicit about bytes vs strings.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1338 @@
+from obspy.core.event import Event, Origin, Magnitude, Comment, Catalog
+from obspy.core.event import EventDescription, CreationInfo
+from obspy.core.event import Pick, WaveformStreamID, Arrival, Amplitude
+import warnings
```
Please change import order (and newlines) to

```python
from __future__ import (absolute_import, division, print_function,
                        unicode_literals)
from future.builtins import *  # NOQA @UnusedWildImport

import warnings

import numpy as np

from obspy import UTCDateTime
from obspy.core.event import Event, Origin, Magnitude, Comment, Catalog
from obspy.core.event import EventDescription, CreationInfo
from obspy.core.event import Pick, WaveformStreamID, Arrival, Amplitude
```

So its future -> builtins -> scipy stack -> obspy

QuLogic replied:
where "scipy stack" is anything 3rd party.

calum-chamberlain commented:
Thanks for those line-notes guys - I will amend the code (thanks for helping me learn!) and work on making sure those tests pass - main issue is is_sfile not coping with being fed io.Bytes.

@QuLogic I will try and remove the string_types in nordic.core too. Thanks!

krischer commented:
Let us know once you think it is ready for a proper review :-)

calum-chamberlain commented:
Looks like it is passing main tests, but failing some coverage tests - I want to remove a couple of extraneous functions (`read_event` is pointless, and a few others are unnecessary in an obspy frame-work).  I plan on leaving in the `readwavename` (might rename with underscores though) and `read_spectral_info`, which won't have direct access from the `read_events` plugin.

Now that it is passing most things (after a lot of issues at my end defining what was required for a nordic s-file, and leaving in print statements when tired) a code-review might be helpful, otherwise, give me a couple of days or so to remove extraneous functions?

calum-chamberlain commented:
@krischer I think it is ready for review, although I'm not sure why codecov/changes is failing, any help on that?

krischer commented:
Trigger doc build: +DOCS

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+def readheader(sfile):
+    """
+    Read header information from a seisan nordic format S-file.
+    Returns an obspy.core.event.Catalog type: note this changed for version \
```
Unnecessary backslash.

QuLogic replied:
Should be a blank line above this one.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    """
+    Read header information from a seisan nordic format S-file.
+    Returns an obspy.core.event.Catalog type: note this changed for version \
+    0.1.0 from the inbuilt class types.
```
What do you mean with "change for version 0.1.0..."?

calum-chamberlain replied:
Ah, hang-over from EQcorrscan, will remove.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    :type f: file
+    :param f: File open in read-mode.
+
+    :returns: obspy.core.event.Event"""
```
Please use 
```
:class:`~obspy.core.event.event.Event`
```

(note the fully qualified name) - then sphinx creates a link in the documentation to the event class.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+
+
+def _readheader(f):
+    """Internal header reader.
```
Please use

```python
def function(...):
    """ (NEWLINE!)
    docstring
    """
    ...
```

everywhere to stay consistent with the rest of ObsPy. You mostly did this everywhere but there are a couple places where you did not.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    new_event = Event()
+    topline = f.readline()
+    if not len(topline.rstrip()) == 80:
+        f.close()
```
No need to close as the function did not open it.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            topline = line
+            break
+        if line[79] == '7':
+            f.close()
```
No need to close.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+                               int(line[18:20]), int(line[20:22]), 0, 0)
+            time += 60  # Add 60 seconds on to the time, this copes with s-file
+            # preference to write seconds in 1-60 rather than 0-59 which
+            # datetime objects accept
```
Could you not solve this through explicit logic - might be cleaner and easier to follow, e.g.

```python
y, m, d, h, min, s, msec = ...
offset = 0
if s == 60:
   s = 0
   offset = 60
time = UTCDateTime(...) + offset
```

krischer replied:
I just saw - you actually do just that a bit further up.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+from obspy.core.event import Pick, WaveformStreamID, Arrival, Amplitude
+
+
+def is_sfile(sfile):
```
Please make this a private function. `_is_sfile(...)`.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        for line in sfile:
+            if line[79] in [' ', '1']:
+                head_line = line
+                break
```
The logic can be  deduplicated by using two functions, e.g.

```python
def _is_sfile(sfile):
    if hasattr(sfile, "readline"):
        return __is_sfile(sfile)
    with open(sfile, "r") as f:
        return __is_sfile(f)

def __is_sfile(sfile):
    Rest of the logic
```

calum-chamberlain replied:
:+1:

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            if line[79] in [' ', '1']:
+                head_line = line
+                break
+    if 'head_line' in locals():
```
Simpler and cleaner:

```python
for line in sfile:
    if ...:
        head_line = line
        break
else:
    return False
```

The `else:` clause is executed if no `break` is encountered. This also avoid a level of indentation which is always nice.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            UTCDateTime(int(head_line[1:5]), int(head_line[6:8]),
+                        int(head_line[8:10]), int(head_line[11:13]),
+                        int(head_line[13:15]), sfile_seconds,
+                        int(head_line[19:20]) * 100000) + add_seconds
```
The logic with the 60 seconds is also needed later on - maybe encapsulate in a function?

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        else:
+            exponant = int('{0:.2E}'.format(number).split('E+')[-1]) - 1
+            divisor = 10 ** exponant
+            string = '{0:.1f}'.format(number / divisor) + 'e' + str(exponant)
```
Can you not just use `"%g" %  number`? This will use exponential notation for larger numbers and normal one for smaller numbers.

calum-chamberlain replied:
Showing my lack of string formatting knowledge here - will do it.

QuLogic replied:
I think this code also changes the `E+` to plain `e`, but this can be done straight away: `('%g' % (number, )).replace('e+', 'e')`

calum-chamberlain replied:
Ah, annoyingly, this also copes with nordic wanting numbers as:
`10.2e5`
rather than:
`1.02e6`
I don't know how to make the formatter work to force that formatting, hence the horrible process here.  I'll take any suggestions though!

krischer replied:
I see. Does Seisan really require that formatting for reading the files or is this just how it tends to spit them out? I just played around a bit but could not find a way to specify the number of digits before the dot using normal string formatting.

calum-chamberlain replied:
Good point, I don't actually know if it requires that format for reading - I will have a play and see.  It definitely spits them out that way though.  From the manual that describes the format they say amplitudes are g7.1, it should cope with e.g. 1.1e6, but it looks like to get more precision you need e.g. 11.2e5 rather than 1.12e6?

krischer replied:
Hmm..do you have a way of knowing what the values are internally read as in seisan? If nothing helps I guess your current solution is alright.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    elif mag_type == 'M':
+        mag = 'W'  # Convert generic magnitude to moment magnitude
+        msg = ('Converting generic magnitude to being stored as moment mag')
+        warnings.warn(msg)
```
This could be simplified with 

```python
mapping = {
    "ML": "L",
    "MS": "S", ...}

mag = mapping[mag_type] if mag_type in mapping else ''
```

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    elif mag_type == 'G':
+        mag = 'MbLg'
+    elif mag_type == 'C':
+        mag = 'Mc'
```
See comment above for a possible simplification.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+                                                int(topline[13:15]),
+                                                sfile_seconds,
+                                                int(topline[19:20]) *
+                                                100000)\
```
See comment above for encapsulating that logic.

megies replied:
I guess @krischer is referring to use of `UTCDateTime.strptime(date_string, format)`..

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        new_event.magnitudes[2].creation_info = \
+            CreationInfo(agency_id=topline[76:79].strip())
+        new_event.magnitudes[2].origin_id = new_event.origins[0].\
+            resource_id
```
Could the last three if conditions be handled in a loop?

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+                           for m in new_event.magnitudes]
+            preferred_magnitude = sorted(_magnitudes,
+                                         key=lambda x: mag_filter.index(x[0]))
+            new_event.preferred_magnitude_id = str(preferred_magnitude[0][1])
```
This should also work:

`new_event.preferred_magnitude_id = preferred_magnitude[0].resource_id`

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            # If there is a magnitude not specified in filter
+            new_event.preferred_magnitude_id =\
+                str(new_event.magnitudes[0].resource_id)
+    elif len(new_event.magnitudes) == 1:
```
This if condition is not needed - this would be handled fine by the previous condition I think.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+                str(new_event.magnitudes[0].resource_id)
+    elif len(new_event.magnitudes) == 1:
+        new_event.preferred_magnitude_id =\
+            str(new_event.magnitudes[0].resource_id)
```
No need for the `str()` (I think).

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    :type sfile: str
+    :param sfile: Sfile to read from.
+
+    :returns: dictionary of spectral information, units as in seisan manual, \
```
Backslash is not needed.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+
+
+def _read_spectral_info(f):
+    """Internal spectral reader.
```
Newline after the three quotes.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    :type f: file
+    :param f: File open in read mode.
+
+    :returns: dictionary of spectral information, units as in seisan manual, \
```
no backslash please.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        if str(line[1:5]) == str('SPEC'):
+            relevant_lines.append(line)
+    spec_inf = []
+    if len(relevant_lines) == 0:
```
More pythonic: `if not relevant_lines: ...`

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        return spec_inf
+    for line in relevant_lines:
+        spec_str = str(line.rstrip().lstrip())
+        if str(spec_str[5:12]) == str('AVERAGE'):
```
No need to convert the string again.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        spec_str = str(line.rstrip().lstrip())
+        if str(spec_str[5:12]) == str('AVERAGE'):
+            info = {}
+            info['station'] = str('AVERAGE')
```
already a string.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        if str(spec_str[5:12]) == str('AVERAGE'):
+            info = {}
+            info['station'] = str('AVERAGE')
+            info['channel'] = str('')
```
already a string.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            info['channel'] = str('')
+            info['moment'] = _float_conv(spec_str[16:22])
+            if info['moment'] == 999:
+                info['moment'] = np.nan
```
NaNs are dangerous - maybe just set it `None`?

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            info['moment_mag'] = _float_conv(spec_str[72:78])
+            spec_inf.append(info)
+            continue
+        if str(spec_str[5:7]) == str('SD'):
```
already a string - also in a bunch of other places in the close vicinity.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    :return: catalog of events
+    :rtype: obspy.core.event.Catalog
+    """
+    from tempfile import NamedTemporaryFile
```
Please try not to use actual temporary files. Consider refactoring the logic to for example use `io.BytesIO` (should be very simple) or maybe just pass the raw file pointer to the underlying functions?

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+
+
+def _read_picks(f, wav_names, new_event):
+    """Internal pick reader. Use a public function instead.
```
Newline after the quotes.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    :param f: File open in read mode
+    :type wav_names: list
+    :param wav_names: List of waveform files in the sfile
+    :type new_event: obspy.core.Event
```
Please use 
```
:class:`~obspy.core.event.event.Event`
```

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            st = obsread(wavefile)
+            evtime = st[0].stats.starttime
+        except:
+            raise IOError('Wavefile: ' + wavefile +
```
Please don't use `IOError` - this error is reserved for Python when an I/O operation like `open()` fails. Use either `ValueError` or just define your own exception, e.g. `NordicSFileParsingError` or something along these lines.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    :type userid: str
+    :param userid: 4-character SEISAN USER ID
+    :type outdir: str
+    :param outdir: Location to write S-file
```
No need for this parameter. Just expect the user to pass the full filename as the `wavefile` parameter.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    :param outdir: Location to write S-file
+    :type overwrite: bool
+    :param overwrite: Overwrite an existing S-file, default=False
+    :type evtime: obspy.core.utcdatetime.UTCDateTime
```
```
:class:`~obspy.core.utcdatetime.UTCDateTime`
```

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+
+    from obspy import read as obsread
+    import os
+    import datetime
```
Please move import to the top of the file except if there is a very good reason not to.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+                str(evtime.second).zfill(2) +
+                'I'.rjust(6) + '\n'))
+    # Write line 3 of s-file
+    write_wavfile = wavefile.split(os.sep)[-1]
```
Just use `os.path.baseline` - also works with weird paths.

QuLogic replied:
`basename`?

krischer replied:
Yep!

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    return sfile
+
+
+def write_select(catalog, filename='select.out', userid='OBSP', evtype='L',
```
Please don't use a default value for the `filename` argument.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+                 wavefiles=None):
+    """Function to write a catalog to a select file in nordic format.
+
+    :type catalog: obspy.core.event.Catalog
```
```
:class:`~obspy.core.event.catalog.Catalog`
```

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    with open(filename, 'w') as fout:
+        for event, wavfile in zip(catalog, wavefiles):
+            sfile = write_nordic(event=event, userid=userid, evtype=evtype,
+                                 wavefiles=wavfile)
```
Please try to work around the temporary files if possible.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    """
+    Write an obspy.event to a nordic formatted s-file.
+
+    :type event: obspy.core.event.Event
```
See comments above.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    :param event: A single obspy event
+    :type filename: str
+    :param filename: Filname to write to, can be None, and filename will be \
+        generated from the origin time in seisan format.
```
Please require users to just pass a filename. This might result in hard to debug errors if this generated random file names.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    :param evtype: Single character string to describe the event, either L, R \
+        or D.
+    :type outdir: str
+    :param outdir: Path to directory to write to
```
See above.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    :type wavefiles: list
+    :param wavefiles: Waveforms to associate the sfile with
+    :type explosion: bool
+    :param explosion: Note if the event is an explosion, will be marked by an \
```
backslash not needed.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    :param explosion: Note if the event is an explosion, will be marked by an \
+        E.
+    :type overwrite: bool
+    :param overwrite: force to overwrite old files, defaults to False
```
Just always overwrite and push that responsibility to the user.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        path, or by looking for the file recursively in directories within \
+        the WAV directory in your seisan install.  Because all lines need to \
+        be less than 79 characters long (fortran hangover) in the s-files, \
+        you will need to determine whether the full-path is okay or not.
```
backslashes not needed here.

megies replied:
:+1:

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    """
+    import datetime
+    import os
+    from obspy.core.event import Catalog, Event
```
Move to top of file.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    else:
+        sfile = open(outdir + os.sep + filename, 'w')
+    # Write the header info.
+    if event.origins[0].latitude:
```
Please do `if event.origins[0].latitude`, otherwise it will also fail if the latitude is 0 which might well be.

QuLogic replied:
@krischer You appear to have written exactly what's already there? I guess you mean `if event.origins[0].latitude is not None:`.

krischer replied:
I guess I should not review stuff late at night...you are of course right.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            lat = ''
+    else:
+        lat = ''
+    if event.origins[0].longitude:
```
Same comment as above.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    return pick_strings
+
+
+def stationtoseisan(station):
```
Could this be hooked into the `obspy.read_inventory()` function?

calum-chamberlain replied:
Good point, I will have a think - there is not much information in there, but it could be used - this is more output from a `Station` class though - might look at adding it as a plugin for `Inventory.write`.

krischer commented:
> although I'm not sure why codecov/changes is failing

Ignore the codecov patch thingy - for some reason it sometimes (quite often...) has coverage changes in unrelated files.

calum-chamberlain commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        except ValueError:
+            # If there is a magnitude not specified in filter
+            new_event.preferred_magnitude_id =\
+                str(new_event.magnitudes[0].resource_id)
```
It might be best not to set it if it is happily valid not to - there is no preferred magnitude in the nordic files.

krischer replied:
Sure. Not setting one is perfectly valid.

calum-chamberlain commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        str(evtime.month).zfill(2)
+    # Check is sfile exists
+    if os.path.isfile(sfile) and not overwrite:
+        print('Desired sfile: ' + sfile + ' exists, will not overwrite')
```
This exists to be consistent with Seisan logic - and is only necessary when filename is not given - Nordic files are named according to the origin time, which is accounted for within this code (e.g. if filename=None then the filename will be set according to the origin time).  Most people will be calling this directly from `catalog.write()` I assume, but I would still like to maintain the functionality that allows it to assign the *correct* (in a seisan sense) filename.
The reason for the overwite being in there is that seisan will try not to overwrite nordic files and will change the filename slightly - can I just set the default to `overwrite=True`, and not set a default `filename`? Then the logic from `Catalog.write` would be consistent with the rest of obspy (I think), but you can still access the more seisan-type logic by calling it functions directly?

krischer replied:
Sounds good :-) As long as the `Catalog.write` behaves in line with all other file formats and the overwriting behavior is documented I'm fine with it. Especially if this is something that people coming from Seisan will expect.

calum-chamberlain commented on obspy/io/nordic/tests/test_nordic.py:
```diff
@@ -0,0 +1,684 @@
+                if not pick_1[key] == pick_2[key]:
+                    print('Different pick: %s' % key)
+                    print(pick_1[key])
+                    print(pick_2[key])
```
Ah, yup, sorry I left them in for debugging - I can remove them.

calum-chamberlain commented:
Thanks for all those comments @krischer, they are great - I hadn't dealt with io things until making this pull request, I will definitely work on not using temporary files there (I hated doing it, but I didn't know any better!).
The NaN/999 things, I think I had a good reason for them once, but I can't remember it, None does seem cleaner.  Nordic files can cope without an origin location (they can just house picks and a time, which, when located is the origin-time, when not, is some assigned time for the start of the waveforms), which is where that stems from.  Shall play and try and get rid of them.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+.. note::
+
+    We have not implemented any handling of focal mechanism solutions between
+    the two formats.
```
`Currently does not`...

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            + add_seconds
+    except:
+        f.close()
+        IOError("Couldn't read a date from sfile")
```
This exception is not actually being raised, but as @krischer says above, this section is probably unnecessary.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    new_event.event_descriptions.append(EventDescription())
+    new_event.event_descriptions[0].text = topline[21:23]
+    # new_event.ev_id=topline[22]
+    if not _float_conv(topline[23:30]) == 999:
```
`_float_conv` is not even useful in this case:
```python
try:
    new_event.origins[0].latitude = float(topline[23:30])
    new_event.origins[0].longitude = float(topline[31:38])
    new_event.origins[0].depth = float(topline[39:43]) * 1000
except ValueError:
    new_event.origins[0].latitude = None
    new_event.origins[0].longitude = None
    new_event.origins[0].depth = None
```
is perfectly clear as is, though as noted below, maybe don't even add the origin.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    # new_event.depth_ind = topline[44]
+    # new_event.loc_ind = topline[45]
+    new_event.creation_info = CreationInfo(agency_id=topline[45:48].
+                                           strip())
```
Does this really not fit on a line?

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+                   topline[49:51].strip())
+    new_event.origins[0].comments.append(ksta)
+    # new_event.origins[0].nsta??? = _int_conv(topline[49:51])
+    if not _float_conv(topline[51:55]) == 999:
```
See, `_float_conv` is making you do double  the conversions, whereas a simple `try:/except ValueError: pass` would have been simpler.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        new_event.origins[0].time_errors['Time_Residual_RMS'] = \
+            _float_conv(topline[51:55])
+    # Read in magnitudes if they are there.
+    if len(topline[59].strip()) > 0:
```
This is `not topline[59].isspace()`, or even `topline[59] != ' '` if you know that's the only whitespace option.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            CreationInfo(agency_id=topline[60:63].strip())
+        new_event.magnitudes[0].origin_id = new_event.origins[0].\
+            resource_id
+    if len(topline[67].strip()) > 0:
```
Same comment as above.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            CreationInfo(agency_id=topline[68:71].strip())
+        new_event.magnitudes[1].origin_id = new_event.origins[0].\
+            resource_id
+    if len(topline[75].strip()) > 0:
```
Same comment as above.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            resource_id
+    if len(topline[67].strip()) > 0:
+        new_event.magnitudes.append(Magnitude())
+        new_event.magnitudes[1].mag = _float_conv(topline[64:67])
```
You don't know that the previous condition was true, so you don't know that `1` is the correct index into the list.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            resource_id
+    # convert the nordic notation of magnitude to more general notation
+    for _magnitude in new_event.magnitudes:
+        _magnitude.magnitude_type = _nortoevmag(_magnitude.magnitude_type)
```
I don't know why this couldn't be done on the line in which it was assigned above.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    for _magnitude in new_event.magnitudes:
+        _magnitude.magnitude_type = _nortoevmag(_magnitude.magnitude_type)
+    # Set the useful things like preferred magnitude and preferred origin
+    new_event.preferred_origin_id = str(new_event.origins[0].resource_id)
```
`str` not necessary as noted below.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        try:
+            # Select moment first, then local, then
+            mag_filter = ['MW', 'Mw', 'ML', 'Ml', 'MB', 'Mb',
+                          'MS', 'Ms', 'Mc', 'MC']
```
Is the case really important? Why does `Mc` come before `MC` when it's the other way around for the others?

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            # Select moment first, then local, then
+            mag_filter = ['MW', 'Mw', 'ML', 'Ml', 'MB', 'Mb',
+                          'MS', 'Ms', 'Mc', 'MC']
+            _magnitudes = [(m.magnitude_type, m.resource_id)
```
You should just supply this as the key, and then you don't need the temporary list.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    if len(relevant_lines) == 0:
+        return spec_inf
+    for line in relevant_lines:
+        spec_str = str(line.rstrip().lstrip())
```
`line` is already a string, and can't you just use `.strip()`?

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            info['station'] = str('AVERAGE')
+            info['channel'] = str('')
+            info['moment'] = _float_conv(spec_str[16:22])
+            if info['moment'] == 999:
```
Anywhere there's some condition, the `_float_conv` makes things more complicated than a simple `try:/except ValueError:`.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            info['moment_mag'] = _float_conv(spec_str[72:78])
+            spec_inf.append(info)
+            continue
+        stachans = [str('.'.join([i['station'], i['channel']]))
```
Redundant `str`.

QuLogic replied:
`stachans` would be better as a `set` created outside the loop to which the station/channel is added when it's added to `spec_inf`. Right now this conversion is done every time the loop is done on every entry in the loop all over again.

It's also not really necessary to flatten it into period-separated string; use a tuple `(station, channel)` instead.

QuLogic replied:
Actually, if you're going to go find the station/channel again (which you do on line 445), then `spec_inf` should be a dictionary, keyed by `(station, channel)`, which is basically a `set`+`list`.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            continue
+        stachans = [str('.'.join([i['station'], i['channel']]))
+                    for i in spec_inf]
+        stachan = str('.'.join([spec_str[4:9].strip(),
```
Redundant `str`.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+                                    phase_hint=phase,
+                                    polarity=polarity, time=time))
+        if line[9] == 'I':
+            new_event.picks[pick_index].onset = 'impulsive'
```
All of this `new_event.picks[pick_index]` stuff looks awfully redundant. Why not use a convenience `pick` variable?

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+                                                  waveform_id))
+            if new_event.picks[pick_index].phase_hint == 'IAML':
+                # Amplitude for local magnitude
+                new_event.amplitudes[amplitude_index].type = 'AML'
```
Same thing about the amplitudes.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+                # Default AML unit in seisan is nm (Page 139 of seisan
+                # documentation, version 10.0)
+                new_event.amplitudes[amplitude_index].generic_amplitude /=\
+                    10**9
```
Use `1e9` to fit on one line.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    :returns: List of strings of wave paths
+    :rtype: list
+    """
+    f = open(sfile, 'r')
```
Should use a context manager.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        raise IOError('Event type must be either L, R or D')
+
+    # Generate s-file name in the format dd-hhmm-ss[L,R,D].Syyyymm
+    sfile = outdir + '/' + str(evtime.day).zfill(2) + '-' +\
```
Use `os.path.join`, not `'/'`

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        str(evtime.minute).zfill(2) + '-' +\
+        str(evtime.second).zfill(2) + evtype + '.S' +\
+        str(evtime.year) +\
+        str(evtime.month).zfill(2)
```
You probably should use `evtime.strftime` instead of doing string formatting manually.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+                str(float(evtime.second)).rjust(4) + ' ' +
+                evtype + '1'.rjust(58) + '\n'))
+    # Write line 2 of s-file
+    f.write(str(' ACTION:ARG ' + str(datetime.datetime.now().year)[2:4] + '-' +
```
You should call `datetime.datetime.now()` once, or there is a very small chance that you will get two different "now"s which will only show up once in a blue moon exactly while running a 30-hour script that you need to have finished running yesterday.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+                'all exist, you need to clean your stuff up!'
+            raise IOError(msg)
+        # sys.exit()
+    f = open(sfile, 'w')
```
Should be in a context manager.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        # sys.exit()
+    f = open(sfile, 'w')
+    # Write line 1 of s-file
+    f.write(str(' ' + str(evtime.year) + ' ' +
```
The outer `str` is redundant; can use `evtime.strftime` to simplify a bit.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    elif isinstance(event, Event):
+        event = event
+    else:
+        raise IOError('Needs a single event')
```
This is a `ValueError`, not an `IOError`.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+                      outdir)
+    # Check that evtype is one of L,R,D
+    if evtype not in ['L', 'R', 'D']:
+        raise IOError('Event type must be either L, R or D')
```
This is a `ValueError`, not an `IOError`.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    try:
+        evtime = event.origins[0].time
+    except IndexError:
+        msg = 'list index out of range: Need at least one origin with at ' +\
```
Everything up to the colon is not needed in this message.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            raise IOError(outdir + os.sep + sfilename +
+                          ' already exists, will not overwrite')
+    else:
+        sfile = open(outdir + os.sep + filename, 'w')
```
Should use a context manager.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+            sfilename = (evtime + add_secs).datetime.strftime('%d-%H%M-%S') +\
+                evtype[0] + '.S' + (evtime + add_secs).\
+                datetime.strftime('%Y%m')
+            if not os.path.isfile(outdir + os.sep + sfilename):
```
Use `os.path.join`, not `os.sep`.

calum-chamberlain commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    return floatstring
+
+
+def _str_conv(number, rounded=False):
```
I'm pretty sure the main reason for this is to cope with the fact that the nordic files can have a lot of uninhabited sections, e.g. nordic files do not need to have lat and lon for them to be valid, amongst a lot of other things.  It was just convenience for the case of converting empty strings to something useful mostly.  I have changed it to remove the `999` and `NaN` values, and put `None` in their place.  I'm pretty sure I could do it inline, but I don't want to repeat it for every element of the nordic file...

QuLogic replied:
OK, I was ignoring the `999` sections, but if they're still relevant with `None`, then maybe this function is useful there. I still think it should _only_ be used where values are optional, otherwise things "work" but are really invalid.

calum-chamberlain replied:
I think (...brave words...) that I have used only where values are optional - but I will have another check.  A lot of things can be missed in nordic files for them still to be valid as far as I have found.  Having looked at the changes I have made for the 999 and nan things I think I could probably do it as well with `try: except:` etc, but they would be almost everywhere.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+        mag_1 = ''
+        mag_1_type = ''
+        mag_1_agency = ''
+    try:
```
These appear to be all the same; you should be able to loop through them as in the read case.

calum-chamberlain commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    else:
+        warnings.warn(mag_type + ' is not convertible')
+        return str('')
+    return str(mag)
```
This might be my ignorance - in my doctests I'm expecting a `str` returned, but in python 3.x I get unicode returned, so they fail - all these redundant `str` commands are coping with that - how should I be doing it / have I created an issue when one isn't there? (Thanks for help in advance!)

QuLogic replied:
There is no unicode in Python 3; it's all `str`.

QuLogic replied:
Maybe you're getting unicode in Python 2 though (the 'u' prefix)? For that the simplest fix is usually to `print` the output.

calum-chamberlain replied:
That'd be it, long week. Thanks, I will make the doctests `print` and remove the redundant `str`s. Thanks!

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    # Work out how many stations were used
+    if len(event.picks) > 0:
+        stations = [pick.waveform_id.station_code for pick in event.picks]
+        ksta = str(len(list(set(stations))))
```
`len` works on `set`; the `list` is unnecessary.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+
+def populatesfile(sfile, event):
+    """
+    Populate a blank nordic format S-file with pick information.
```
Blank line after this.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+def populatesfile(sfile, event):
+    """
+    Populate a blank nordic format S-file with pick information.
+    arguments required are the filename of the blank s-file and the picks \
```
This paragraph should probably go in the description of the `picks` parameter below. It doesn't seem to be about the function in general.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    return str(sfilename)
+
+
+def populatesfile(sfile, event):
```
`populate_sfile`.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    elif isinstance(event, Event):
+        event = event
+    else:
+        raise AttributeError('More than one event in the catalog, use a ' +
```
I think that's a `ValueError`, not an `AttributeError`.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    else:
+        raise AttributeError('More than one event in the catalog, use a ' +
+                             ' different method')
+    f = open(sfile, 'r')
```
Should use a context manager.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    # picks there we should preserve them
+    body = ''
+    header = ''
+    if 'headerend' in locals():
```
Avoid using `locals`.

QuLogic commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    # Now generate lines for the new picks
+    newpicks = '\n'.join(nordpick(event))
+    # Write all new and old info back in
+    f = open(sfile, 'w')
```
Should use a context manager.

QuLogic commented on obspy/io/nordic/tests/test_nordic.py:
```diff
@@ -0,0 +1,684 @@
+        test_cat = Catalog()
+        test_cat += test_event
+        # Write the catalog
+        test_cat.write("Test_catalog.xml", format='QUAKEML')
```
This does not appear to be necessary.

QuLogic commented on obspy/io/nordic/tests/test_nordic.py:
```diff
@@ -0,0 +1,684 @@
+        read_cat = read_events("Test_catalog.xml")
+        os.remove("Test_catalog.xml")
+        self.assertEqual(read_cat[0].resource_id, test_cat[0].resource_id)
+        for i in range(len(read_cat[0].picks)):
```
```python
self.assertEqual(len(read_cat[0].picks), len(test_cat[0].picks))
for read_pick, test_pick in zip(read_cat[0].picks, test_cat[0].picks):
    ...
```

QuLogic commented on obspy/io/nordic/tests/test_nordic.py:
```diff
@@ -0,0 +1,684 @@
+                    continue
+                self.assertEqual(read_cat[0].picks[i][key],
+                                 test_cat[0].picks[i][key])
+        self.assertEqual(read_cat[0].origins[0].resource_id,
```
Maybe you should set `read_ev = read_cat[0]` and `test_ev = test_cat[0]`, and then you don't need to use all those `[0]`.

QuLogic commented on obspy/io/nordic/tests/test_nordic.py:
```diff
@@ -0,0 +1,684 @@
+        self.assertEqual(readwavename(sfile), ['test'])
+        read_cat = Catalog()
+        read_cat += read_nordic(sfile)
+        os.remove(sfile)
```
Anything that needs this should be in a
```python
try:
    ...
finally:
    os.remove(sfile)
```

QuLogic commented on obspy/io/nordic/tests/test_nordic.py:
```diff
@@ -0,0 +1,684 @@
+                         test_cat[0].amplitudes[1].category)
+        del read_cat
+
+        # Test a deliberate fail
```
Seems like this is really a different and separate test case.

QuLogic commented on obspy/io/nordic/tests/test_nordic.py:
```diff
@@ -0,0 +1,684 @@
+        self.assertEqual(read_cat[0].amplitudes[1].category,
+                         test_cat[0].amplitudes[1].category)
+
+        # Check the read-write s-file functionality
```
This could be a separate test, I think.

QuLogic commented on obspy/io/nordic/tests/test_nordic.py:
```diff
@@ -0,0 +1,684 @@
+                         wavefiles='test', explosion=True,
+                         overwrite=True)
+        with self.assertRaises(IndexError):
+            invalid_origin = test_cat[0].copy()
```
This probably shouldn't be in the context manager.

QuLogic commented on obspy/io/nordic/tests/test_nordic.py:
```diff
@@ -0,0 +1,684 @@
+                         overwrite=True)
+        with self.assertRaises(ValueError):
+            invalid_origin = test_cat[0].copy()
+            invalid_origin.origins[0].time = None
```
This probably shouldn't be in the context manager either.

calum-chamberlain commented on obspy/io/nordic/core.py:
```diff
@@ -0,0 +1,1379 @@
+    for lineno, line in enumerate(f):
+        if 'headerend' in locals():
+            if len(line.rstrip('\n').rstrip('\r')) in [80, 79] and \
+               (line[79] == ' ' or line[79] == '4' or line[79] == '\n'):
```
:+1:

calum-chamberlain commented:
Last few commits try to address everything, although I haven't hooked station_to_seisan into inventory.write at the moment.  Mostly I forgot before I pushed.  If you guys ant me to try and do that tomorrow I can have a play - shouldn't be too difficult.  There wouldn't be an equivalent read_inventory for it though, there is not much information in the seisan STATION0.HYP file, it's just station co-ordinates used for locations.
Let me know what you think (assuming the tests pass, passed on my python 2.7 and 3.5 here, but there will probably be something that I have missed!).

calum-chamberlain commented:
Looks like just flake8 errors - for some reason all my python environments give different flake8 errors (including when I set-up a test environment following the .travis file) to the CI bots.  Must be doing something wrong - if you guys have any suggestions for that so I can replicate the flake8 test environment locally and not have these errors would be much appreciated.

I will fix those errors tomorrow NZ time and push. If you see anything else that needs doing let me know.

Apologies for the last commit as well - I should have rebased, but it was late in the day and I wasn't thinking!

Thanks for those reviews, definitely made it a lot better (a lot of fluff had been accumulated from learning python, and patching holes).

calum-chamberlain commented:
Added a plugin for Inventory.write as a SEISAN format (and a test), and fixed the flake8 errors.  From my end I think it might be there.
I haven't made use of str.format, but I have removed a **lot** of temporary variables that were a hang-over from writing the module before integrating it with Event classes.
Let me know if you need anything more from me.

megies commented on obspy/CONTRIBUTORS.txt:
```diff
@@ -9,6 +9,7 @@ Felix Bernauer
 Moritz Beyreuther
 Sébastien Bonaimé
 Lloyd Carothers
+Calum Chamberlain
```
this will need a small change if #1533 is merged before..

calum-chamberlain replied:
👍

calum-chamberlain replied:
@megies should I rebase this now then?  I have addressed the comments below now too (when I push I will put some update highlights).

calum-chamberlain replied:
Rebased, running tests locally, will push in a jiffy.

megies commented on setup.py:
```diff
@@ -356,6 +363,9 @@
         'writeFormat = obspy.io.css.station:_write_css',
         ],
+    'obspy.plugin.inventory.SEISAN': [
+        'writeFormat = obspy.io.nordic.core:inventory_to_seisan',
```
Hmm.. we also already have an `obspy.io.seisan` submodule (for waveforms), so maybe it's worth a thought to move the SEISAN output part there?

calum-chamberlain replied:
Mmm, I had a thought about that but forgot to mention it - happy to move it there - it's a slightly odd case in that it doesn't fit with the event things in `obspy.io.nordic`, nor really with the waveform things in `obspy.io.seisan`.  The format is probably more correctly called *STATION0.HYP* which is the file that seisan uses as input for *Hypocentre*.  I think either a name change to:
```python
   'obspy.plugin.inventory.STATION0': [...
```
or moving to the `obspy.io.seisan` would be worthwhile - any preference or other ideas?

calum-chamberlain replied:
This PR is solely about Nordic format, however, this was a little addition I had in EQcorrscan - it was suggested I could include it and plug it in as above (although naming isn't clear, it isn't a clear format in my mind, and this function doesn't make the whole file, there are other elements to it).  This function was just convenience for me.  Happy to remove and keep to myself?

megies replied:
If it is..
 - neither a well defined data format
 - nor used by a community or some widely used third-party program for I/O,
 - nor used by the rest of the PR (nordic file format I/O)..

..then I would say we should drop it.

calum-chamberlain replied:
👍 Have moved this from here and made it a gist, so if anyone comes looking they can find it [here.](https://gist.github.com/calum-chamberlain/ca84185a324dc42d631808d94e4ad0c6)

megies commented on obspy/io/nordic/tests/test_nordic.py:
```diff
@@ -0,0 +1,698 @@
+                self.assertTrue(test_similarity(event_1=event_1,
+                                                event_2=event_2))
+        finally:
+            os.remove('select.out')
```
For this we have a convenience construct (see [here](http://docs.obspy.org/master/packages/autogen/obspy.core.util.base.NamedTemporaryFile.html?highlight=named#obspy.core.util.base.NamedTemporaryFile)), 
the file will be removed automatically when exiting the context..


That would make this test:

```python
def test_write_select(self):
    cat = read_events()
    with NamedTemporaryFile(suffix='.out') as tf: 
        write_select(cat, filename=tf.name)
        cat_back = read_events(tf.name)
    for event_1, event_2 in zip(cat, cat_back):
        self.assertTrue(test_similarity(event_1=event_1,
                                        event_2=event_2))
```

Would be good if you could adapt your tests..

calum-chamberlain replied:
Ah, I was doing this following a comment above, but I can change again - there are quite a few tests that will have to be updated.

megies replied:
Sorry, if I was cross-firing with another comment. But I think this is the most obspy-canonical handling of temp files in tests. It's always better to do the nitty gritty (i.e. handle the temp file removal) in one place (i.e. inside the implementation of `NamedTemporaryFile`) rather than duplicated across code (which is hard to maintain) and in less detail (`NamedTemporaryFile` checks if the file really exists before trying to remove it.. if something goes wrong in your code and the file isn't even created in the first place, your test will spit out some `IOError` (that needs debugging subsequently..)

calum-chamberlain replied:
I have adapted tests where appropriate, although some are left deleting temporary files. 
Those that do remove temporary files have the filename returned to them on writing (the internal nordic writers can generate automatic filenames, and I want to test that).  I think the only way that the file that the test tries to delete could not-exist is if the earlier statements failed (which would raise a more useful error earlier; this is for tests that were not in a `try: ... finally:` statement, which have all been removed/converted to the `NamedTemporaryFile` syntax).

calum-chamberlain commented on obspy/io/nordic/tests/test_nordic.py:
```diff
@@ -0,0 +1,698 @@
+                                                  time_residual=0.2,
+                                                  distance=15,
+                                                  azimuth=25))
+
```
I wish I had known that was allowed! I will be doing that in future - and here if I get the time. Thanks!

calum-chamberlain replied:
:+1: Done

calum-chamberlain commented:
Any help with the docker-testbot fail would be great - doesn't look like an issue from me (fails in obspy.core.  Otherwise I think I have made all the changes requested?

megies commented:
> Any help with the docker-testbot fail would be great - doesn't look like an issue from me

Can be ignored, that test fail is unrelated and fixed in master already.

But.. @calum-chamberlain  it seems the history of your branch is messed up.. there's tons of unrelated commits in here, I suspect some rebase that went wrong is responsible?

calum-chamberlain commented:
It is incredibly messy, and it is definitely the rebase. Any way to get around that?

QuLogic commented:
You can fix it with an interactive rebase. I tried to do it and got 90% of the way there, except I haven't been paying attention to this PR, so I wasn't sure how to fix the last few conflicts.

```
$ git fetch upstream
$ git checkout master
$ git merge upstream/master
$ git checkout feature/nordic_integration
$ git branch feature/original_nordic_integration  # Backup branch
$ git rebase -i master
```
Then you should get a list of commits to include. There are basically two copies of _almost_ everything.

The first 22 commits can be removed entirely as they repeat later.

I am not sure of the next 3 commits (from "Remove inventory-to-station: now gist" to "flake8"); maybe they can be dropped? If not dropping them, then move those lines after "Add SEISAN write plugin for inventory".

The next 22 commits are copies of the first set and should be kept, since they're the more recent version.

The last 2 commits should be kept at the very end.

This generally works except for a couple of expected conflicts with changelog and contributor file. Then at nearly the very end, there's one last conflict that I'm not sure how to fix. I think it's because you mixed the merge commits with actual flake8 changes. It's not a good idea to do extra work in merge commits. You might be able to do `git rebase --preserve-merges -i master` and see if including the merges will let you find the remaining changes, but I can't be certain how well that will work.

If in the end, everything worked perfectly, then force push the result here: `git push --force upstream feature/nordic_integration`. If something went horribly wrong, `git rebase --abort` (if necessary) and `git reset --hard feature/original_nordic_integration` and try again.

QuLogic commented:
Wow, actually, you can scratch most of that stuff by using `--preserve-merges`. It enables git to figure out that half the commits are duplicates and you don't even need to worry about them.

```
$ git fetch upstream  # Assuming that's here.
$ git checkout master
$ git merge upstream/master
$ git checkout feature/nordic_integration
$ git branch feature/original_nordic_integration  # Backup branch
$ git rebase --preserve-merges -i master
```

The list will then have only _one_ copy of the duplicate commits, and all you have to do is fix the conflicts at the (old) merge commit.

If in the end, everything worked perfectly, then force push the result here: `git push --force upstream feature/nordic_integration`. If something went horribly wrong, `git rebase --abort` (if necessary) and `git reset --hard feature/original_nordic_integration` and try again.

calum-chamberlain commented:
@QuLogic Thanks for the help! I'm afraid I need more advice though - with the CONTRIBUTORS.txt file, at every patch attempt the rebase tries and fails to merge that, what should I do for that file (and the CHANGES.md file)?

QuLogic commented:
You'll need to fix the conflicts manually; should be something like:
```
other stuff without conflict
<<<<<< HEAD
stuff on master
======
stuff on rebased commit
>>>>>> commit 123456
other stuff without conflict
```
You'll need to figure out what part to drop yourself  (e.g., `CHANGELOG.txt` should have the `obspy.nordic` stuff inserted after `obspy.nlloc` and for `CONTRIBUTORS.txt`, your name will have to be inserted in the right place, etc.)

Then `git add conflicting-file` and `git rebase --continue`.

calum-chamberlain commented:
@QuLogic Thanks! Sorry, I should have just done that, I was concerned that it was that part that had mucked up the history, obviously it was all the extra, duplicate commits.  This looks better I think? Just check that I haven't somehow broken anything (tests run locally). Thanks and sorry this has been so painful guys.

QuLogic commented:
Sorry, looks like you've still got duplicate commits somehow. Maybe it was an older version of git that isn't as smart?

calum-chamberlain commented:
Ah, I do!  That's really strange, git popped up a to-to list before the rebase got fully underway and there weren't any duplicates. That is really frustrating.  What version of git are you using @QuLogic (I have 2.7.4 here)?

megies commented:
> Wow, actually, you can scratch most of that stuff by using --preserve-merges. It enables git to figure out that half the commits are duplicates and you don't even need to worry about them.

Interesting.. good to know!

megies commented:
I just had a look and tried to rebase, but I'm running into tons of merge conflicts while trying to clean up. I fear the history is beyond repair for people that have not worked on this themselves (unless somebody wants to spend some hours on it), and I would vote to just make this a single commit from the current state.

I pushed a single commit version here: https://github.com/obspy/obspy/tree/nordic_take_2

If the history of the individual commits is worth keeping, we can start another pull request and leave this PR as is unmerged. Otherwise we can overwrite the branch of this PR with the single commit branch. I don't particularly care either way, so I'll leave it to @calum-chamberlain to decide.

calum-chamberlain commented:
I vote squash into a single commit, I don't think the history matters in this case.

Sent from my Vodafone Smart

On Nov 8, 2016 11:54 PM, Tobias Megies <notifications@github.com> wrote:

I just had a look and tried to rebase, but I'm running into tons of merge conflicts while trying to clean up. I fear the history is beyond repair for people that have not worked on this themselves (unless somebody wants to spend some hours on it), and I would vote to just make this a single commit from the current state.

I pushed a single commit version here: https://github.com/obspy/obspy/tree/nordic_take_2

If the history of the individual commits is worth keeping, we can start another pull request and leave this PR as is unmerged. Otherwise we can overwrite the branch of this PR with the single commit branch. I don't particularly care either way, so I'll leave it to @calum-chamberlain<https://github.com/calum-chamberlain> to decide.

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/obspy/obspy/pull/1517#issuecomment-259106598>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKaGc2t1odIl6v26IWG39rxMM2vqrKyKks5q8FT0gaJpZM4J0urn>.

megies commented on obspy/CONTRIBUTORS.txt:
```diff
@@ -70,3 +71,4 @@ Wassermann, Joachim
 Williams, Mark C.
 Winkelman, Andrew
 Zad, Seyed Kasra Hosseini
+
```
Docs build is currently failing due to this empty line..

megies commented:
Alright, I'll force-push to this branch then. For future reference, the original history of this PR can still be looked up by looking at commit e0931322f8806df79352f6d2e1dd7f1674896751.

megies commented:
Another rebase+force-push to get rid of (trivial) merge conflicts.

megies commented:
Is there anything left to address in this PR @calum-chamberlain? Looks like this should be ready now?

calum-chamberlain commented:
Should be good to go I think.

Sent from my Vodafone Smart

On Nov 16, 2016 1:38 AM, Tobias Megies <notifications@github.com> wrote:

Is there anything left to address in this PR @calum-chamberlain<https://github.com/calum-chamberlain>? Looks like this should be ready now?

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/obspy/obspy/pull/1517#issuecomment-260630388>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKaGczSSXXrdyiK4r1fH99YJ9mjDlgBtks5q-ae2gaJpZM4J0urn>.

trichter commented on setup.py:
```diff
@@ -90,11 +90,11 @@
+    'NERIES', 'NonLinLoc', 'NLLOC', 'Nordic', 'observatory', 'ORFEUS', 'PDAS',
+    'picker', 'processing', 'PQLX', 'Q', 'real time', 'realtime', 'REFTEK',
+    'REFTEK130', 'RT-130', 'RESP', 'response file', 'RT', 'SAC', 'sc3ml',
+    'SDS', 'SEED', 'SeedLink', 'SEG-2', 'SEG Y', 'SEISAN', 'SeisHub', 'Seismic
```
A line break may not be within a string. Better break after `'SeisHub',`.

calum-chamberlain replied:
Agreed, thanks.

megies commented:
Gonna merge this now, the only review comments that weren't addressed yet are mostly cosmetics, I think.
